index_documents.go

```diff
@@ -366,6 +370,169 @@ func (i Index) UpdateDocumentsInBatches(documentsPtr interface{}, batchSize int,
 	return resp, nil
 }
 
+func (i Index) UpdateDocumentsCsv(documents []byte, primaryKey ...string) (resp *TaskInfo, err error) {
+	return i.updateDocuments(documents, contentTypeJSON, primaryKey...)
+}
+
+func (i Index) UpdateDocumentsCsvInBatches(documents []byte, batchSize int, primaryKey ...string) (resp []TaskInfo, err error) {
+	// Reuse io.Reader implementation
+	return i.updateDocumentsCsvFromReaderInBatches(bytes.NewReader(documents), batchSize, primaryKey...)
+}
+
+func (i Index) updateDocumentsCsvFromReaderInBatches(documents io.Reader, batchSize int, primaryKey ...string) (resp []TaskInfo, err error) {
+	// Because of the possibility of multiline fields it's not safe to split
+	// into batches by lines, we'll have to parse the file and reassemble it
+	// into smaller parts. RFC 4180 compliant input with a header row is
+	// expected.
+	// Records are read and sent continuously to avoid reading all content
+	// into memory. However, this means that only part of the documents might
+	// be added successfully.
+
+	var (
+		responses []TaskInfo
+		header    []string
+		records   [][]string
+	)
+
+	sendCsvRecords := func(records [][]string) (*TaskInfo, error) {
+		b := new(bytes.Buffer)
+		w := csv.NewWriter(b)
+		w.UseCRLF = true // Keep output RFC 4180 compliant
+		err := w.WriteAll(records)
+		if err != nil {
+			return nil, fmt.Errorf("could not write CSV records: %w", err)
+		}
+
+		resp, err := i.UpdateDocumentsCsv(b.Bytes(), primaryKey...)
```

Comment:
Except for this line, this method is the exact same as `AddDocumentsNdjsonFromReaderInBatches` maybe we can find a way to re-used the code that already existing than duplicated it? WDYT?

Reply:
Sure, I think a higher order function will resolve this.

Reply:
> Except for this line, this method is the exact same as `AddDocumentsNdjsonFromReaderInBatches` maybe we can find a way to re-used the code that already existing than duplicated it? WDYT?

I hope you mean `AddDocumentsCsvFromReaderInBatches`